lnrpc/invoicesrpc/addinvoice.go

```diff
@@ -111,6 +116,42 @@ type AddInvoiceData struct {
 // paymentHashAndPreimage returns the payment hash and preimage for this invoice
 // depending on the configuration.
 func (d *AddInvoiceData) paymentHashAndPreimage() (*lntypes.Preimage, lntypes.Hash, error) {
+	if d.Amp {
+		return d.ampPaymentHashAndPreimage()
+	}
+
+	return d.mppPaymentHashAndPreimage()
+}
+
+// ampPaymentHashAndPreimage returns the payment hash to use for an AMP invoice.
+// The preimage will always be nil.
+func (d *AddInvoiceData) ampPaymentHashAndPreimage() (*lntypes.Preimage, lntypes.Hash, error) {
+	switch {
+
+	// Preimages cannot be set on AMP invoice.
+	case d.Preimage != nil:
+		return nil, lntypes.Hash{},
+			errors.New("preimage set on AMP invoice")
+
+	// If a specific hash was requested, use that.
+	case d.Hash != nil:
+		return nil, *d.Hash, nil
+
+	// Otherwise generate a random hash value, just needs to be unique to be
+	// added to the invoice index.
```

Comment:
Does this mean that the hash is only used for the internal purpose of fitting it into an existing index?

Reply:
Yep, so we can't remove the payment hash without also breaking invoice parsing for other clients. And rather than returning parsing error, it is preferable to use the invoice feature bits to signal incompatibilities, that way at least there is an actionable error message (or at least that was the conclusion we arrived at when adding invoice feature bits in the first place)